### PR TITLE
[IDEA] Add KotlinClassConstructorInfoHandler for showing parameter info of parameterized super class constructor.

### PR DESCRIPTION
**Pull request by pocmo**
Fixes [KT-41617](https://youtrack.jetbrains.com/issue/KT-41617).

### CODE REVIEW

File: idea/src/org/jetbrains/kotlin/idea/parameterInfo/KotlinTypeArgumentInfoHandler.kt [hunk 1]
```diff
@@ -40,12 +37,33 @@ import org.jetbrains.kotlin.types.typeUtil.TypeNullability
 import org.jetbrains.kotlin.types.typeUtil.isAnyOrNullableAny
 import org.jetbrains.kotlin.types.typeUtil.nullability
 
+class KotlinClassConstructorInfoHandler : KotlinTypeArgumentInfoHandlerBase<ClassConstructorDescriptor>() {
+    override fun fetchTypeParameters(descriptor: ClassConstructorDescriptor): List<TypeParameterDescriptor> = descriptor.typeParameters
+
+    override fun findParameterOwners(argumentList: KtTypeArgumentList): Collection<ClassConstructorDescriptor>? {
+        val userType = argumentList.parent as? KtUserType ?: return null
+        val descriptors = userType.referenceExpression?.resolveMainReferenceToDescriptors()?.mapNotNull { it as? ClassConstructorDescriptor }
+        return if (descriptors?.isEmpty() == true) {
```
Comment (vladimirdolzhenko) [inline idea/src/org/jetbrains/kotlin/idea/parameterInfo/KotlinTypeArgumentInfoHandler.kt:46]: `descriptors?.takeIf { it.isNotEmpty() }` looks better

File: idea/src/org/jetbrains/kotlin/idea/parameterInfo/KotlinTypeArgumentInfoHandler.kt [hunk 2]
```diff
@@ -40,12 +37,33 @@ import org.jetbrains.kotlin.types.typeUtil.TypeNullability
...
+}
+
 class KotlinClassTypeArgumentInfoHandler : KotlinTypeArgumentInfoHandlerBase<ClassDescriptor>() {
     override fun fetchTypeParameters(descriptor: ClassDescriptor) = descriptor.typeConstructor.parameters
 
     override fun findParameterOwners(argumentList: KtTypeArgumentList): Collection<ClassDescriptor>? {
         val userType = argumentList.parent as? KtUserType ?: return null
-        return userType.referenceExpression?.resolveMainReferenceToDescriptors()?.mapNotNull { it as? ClassDescriptor }
+        val descriptors = userType.referenceExpression?.resolveMainReferenceToDescriptors()?.mapNotNull { it as? ClassDescriptor }
+        return if (descriptors?.isEmpty() == true) {
```
Comment (vladimirdolzhenko) [inline idea/src/org/jetbrains/kotlin/idea/parameterInfo/KotlinTypeArgumentInfoHandler.kt:62]: same here for `takeIf`

File: idea/src/org/jetbrains/kotlin/idea/parameterInfo/KotlinTypeArgumentInfoHandler.kt (hunk 1 above)
Comment (vladimirdolzhenko) [reply idea/src/org/jetbrains/kotlin/idea/parameterInfo/KotlinTypeArgumentInfoHandler.kt:46]: so - lines 45-46 could be merged into a single one

**Comment by pocmo** [comment]
@vladimirdolzhenko Thank you for the quick review. Agreed, updated to the shorter version with `takeIf`. :)

**Comment by vladimirdolzhenko** [comment]
Small note: it is not necessary to force push your commits - it makes review even more difficult (but this PR is a small one and that's ok), 
Anyway we usually rebase and squash all commits.

LGTM, Waiting for CI

File: idea/testData/parameterInfo/typeArguments/ParameterizedClassConstructor.kt [hunk 3]
```diff
@@ -0,0 +1,13 @@
+open class Parameterized<A, C: Cloneable, R: Runnable>
+
+class AC
+class CC : Cloneable
+class RC : Runnable {
+    override fun run() {}
+}
+
+fun use(p: Parameterized<AC, CC, RC>) {}
```
Comment (vladimirdolzhenko) [inline idea/testData/parameterInfo/typeArguments/ParameterizedClassConstructor.kt:9]: what is the purpose of this `use` function ?

**Comment by pocmo** [comment]
> Small note: it is not necessary to force push your commits - it makes review even more difficult (but this PR is a small one and that's ok),

Got it. 👍 

> LGTM, Waiting for CI

Is this accessible publicly somewhere? Since nothing seems to get reported back to GitHub from CI.

**Comment by vladimirdolzhenko** [comment]
Merged manually

**Comment by vladimirdolzhenko** [comment]
> Is this accessible publicly somewhere? Since nothing seems to get reported back to GitHub from CI.

No, it is private CI, we had public CI in past - maybe we can reconsider to have it public again.

**Comment by pocmo** [comment]
Thank you!

In case you know any other bugs in that area that are great candidates for new contributors then please let me know. Would definitely like to contribute more and explore the code base more. Otherwise I may skim the bug tracker a bit tonight. :)

**Pull request closed without merging**